File: src/lib/workflow.ts
import { Document } from '@/types/document'
import { VerificationResult } from '@/lib/verification'

export type WorkflowState = 'planning' | 'executing' | 'verifying' | 'completed' | 'failed'

export interface Task {
  id: string
  title: string
  description: string
  status: 'pending' | 'in-progress' | 'completed' | 'failed'
  documentId?: string
  verificationResultId?: string
  dependencies: string[]
  createdAt: string
  completedAt?: string
}

export interface Workflow {
  id: string
  title: string
  state: WorkflowState
  tasks: Task[]
  currentTaskId?: string
  documents: Document[]
  verificationResults: VerificationResult[]
  createdAt: string
  updatedAt: string
}

export function createWorkflow(title: string): Workflow {
  const id = Date.now().toString()
  const now = new Date().toISOString()
  
  return {
    id,
    title,
    state: 'planning',
    tasks: [],
    documents: [],
    verificationResults: [],
    createdAt: now,
    updatedAt: now
  }
}

export function addTask(
  workflow: Workflow,
  task: Omit<Task, 'id' | 'createdAt' | 'status' | 'dependencies'>
): Workflow {
  const newTask: Task = {
    ...task,
    id: Date.now().toString(),
    createdAt: new Date().toISOString(),
    status: 'pending',
    dependencies: task.dependencies || []
  }
  
  return updateWorkflow(workflow, {
    tasks: [...workflow.tasks, newTask]
  })
}

export function updateTask(
  workflow: Workflow,
  taskId: string,
  updates: Partial<Task>
): Workflow {
  const newTasks = workflow.tasks.map(task =>
    task.id === taskId ? { ...task, ...updates } : task
  )
  
  return updateWorkflow(workflow, { tasks: newTasks })
}

export function completeTask(
  workflow: Workflow,
  taskId: string
): Workflow {
  const task = workflow.tasks.find(t => t.id === taskId)
  if (!task) return workflow
  
  return updateTask(workflow, taskId, {
    status: 'completed',
    completedAt: new Date().toISOString()
  })
}

export function updateWorkflow(
  workflow: Workflow,
  updates: Partial<Workflow>
): Workflow {
  return {
    ...workflow,
    ...updates,
    updatedAt: new Date().toISOString()
  }
}

export function getNextTask(workflow: Workflow): Task | null {
  if (workflow.state !== 'executing') return null
  
  const currentTask = workflow.tasks.find(t => t.id === workflow.currentTaskId)
  
  if (currentTask?.status === 'completed') {
    const pendingTasks = workflow.tasks.filter(
      t => t.status === 'pending' && t.dependencies.every(depId => {
        const dep = workflow.tasks.find(t => t.id === depId)
        return dep?.status === 'completed'
      })
    )
    
    if (pendingTasks.length > 0) {
      return pendingTasks[0]
    }
    
    if (workflow.tasks.every(t => t.status === 'completed')) {
      return null
    }
  }
  
  return currentTask || null
}

export function canExecuteTask(task: Task, workflow: Workflow): boolean {
  if (task.status !== 'pending') return false
  
  return task.dependencies.every(depId => {
    const dep = workflow.tasks.find(t => t.id === depId)
    return dep?.status === 'completed'
  })
}

export function startWorkflow(workflow: Workflow): Workflow {
  return updateWorkflow(workflow, {
    state: 'executing',
    currentTaskId: workflow.tasks.find(t => t.status === 'pending')?.id
  })
}

export function completeWorkflow(workflow: Workflow): Workflow {
  const allCompleted = workflow.tasks.every(t => t.status === 'completed')
  
  if (allCompleted) {
    return updateWorkflow(workflow, {
      state: 'completed'
    })
  }
  
  return workflow
}

File: src/lib/workflow.py
"""Workflow types and utilities."""
# -*- coding: utf-8 -*-

from __future__ import annotations

import sys
import os

from models.data import Workflow, Task
from datetime import datetime

# Ensure stdout/stderr are UTF-8
if hasattr(sys.stdout, 'buffer') and sys.stdout.encoding and sys.stdout.encoding.lower() != 'utf-8':
    sys.stdout = open(sys.stdout.fileno(), mode='w', encoding='utf-8', buffering=1)
if hasattr(sys.stderr, 'buffer') and sys.stderr.encoding and sys.stderr.encoding.lower() != 'utf-8':
    sys.stderr = open(sys.stderr.fileno(), mode='w', encoding='utf-8', buffering=1)
os.environ.setdefault('PYTHONIOENCODING', 'utf-8')


def create_workflow(title: str) -> Workflow:
    """Create a new workflow."""
    now = datetime.now().isoformat()
    return Workflow(
        id=str(len([]) + 1),  # Simple ID generation
        title=title,
        tasks=[],
        created_at=now,
        updated_at=now,
    )


def add_task(
    workflow: Workflow,
    task: Task,
) -> Workflow:
    """Add a task to workflow."""
    workflow.tasks.append(task)
    workflow.updated_at = datetime.now().isoformat()
    return workflow


def complete_task(workflow: Workflow, task_id: str) -> Workflow:
    """Mark a task as completed."""
    for task in workflow.tasks:
        if task.id == task_id:
            task.status = "completed"
            task.completed_at = datetime.now().isoformat()
            break
    workflow.updated_at = datetime.now().isoformat()
    return workflow


def start_workflow(workflow: Workflow) -> Workflow:
    """Start workflow execution."""
    workflow.state = "executing"
    pending = next((t for t in workflow.tasks if t.status == "pending"), None)
    if pending:
        workflow.current_task_id = pending.id
    workflow.updated_at = datetime.now().isoformat()
    return workflow


def complete_workflow(workflow: Workflow) -> Workflow:
    """Mark workflow as completed if all tasks are done."""
    if all(t.status == "completed" for t in workflow.tasks):
        workflow.state = "completed"
    workflow.updated_at = datetime.now().isoformat()
    return workflow


def get_next_task(workflow: Workflow) -> Task | None:
    """Get next pending task."""
    if workflow.state != "executing":
        return None

    current = next(
        (t for t in workflow.tasks if t.id == workflow.current_task_id), None
    )
    if current and current.status == "completed":
        pending = [
            t
            for t in workflow.tasks
            if t.status == "pending"
            and all(
                (dep := next((x for x in workflow.tasks if x.id == dep_id), None))
                and dep.status == "completed"
                for dep_id in t.dependencies
            )
        ]
        if pending:
            return pending[0]
        if all(t.status == "completed" for t in workflow.tasks):
            return None
    return current


def can_execute_task(task: Task, workflow: Workflow) -> bool:
    """Check if a task can be executed."""
    if task.status != "pending":
        return False
    return all(
        (dep := next((t for t in workflow.tasks if t.id == dep_id), None))
        and dep.status == "completed"
        for dep_id in task.dependencies
    )
